core/trino-main/src/main/java/io/trino/server/PluginManager.java

```diff
@@ -239,95 +203,24 @@ private void installPluginInternal(Plugin plugin, Supplier<ClassLoader> duplicat
-        return buildClassLoaderFromCoordinates(plugin);
-    }
-
-    private PluginClassLoader buildClassLoaderFromPom(File pomFile)
-            throws Exception
-    {
-        List<Artifact> artifacts = resolver.resolvePom(pomFile);
-        PluginClassLoader classLoader = createClassLoader(artifacts, pomFile.getPath());
-
-        Artifact artifact = artifacts.get(0);
-        Set<String> plugins = discoverPlugins(artifact, classLoader);
-        if (!plugins.isEmpty()) {
-            File root = new File(artifact.getFile().getParentFile().getCanonicalFile(), "plugin-discovery");
-            writePluginServices(plugins, root);
-            log.debug("    %s", root);
-            classLoader = classLoader.withUrl(root.toURI().toURL());
-        }
-
-        return classLoader;
-    }
-
-    private PluginClassLoader buildClassLoaderFromDirectory(File dir)
-            throws Exception
-    {
-        log.debug("Classpath for %s:", dir.getName());
-        List<URL> urls = new ArrayList<>();
-        for (File file : listFiles(dir)) {
-            log.debug("    %s", file);
-            urls.add(file.toURI().toURL());
-        }
-        return createClassLoader(urls);
-    }
-
-    private PluginClassLoader buildClassLoaderFromCoordinates(String coordinates)
-            throws Exception
-    {
-        Artifact rootArtifact = new DefaultArtifact(coordinates);
-        List<Artifact> artifacts = resolver.resolveArtifacts(rootArtifact);
-        return createClassLoader(artifacts, rootArtifact.toString());
-    }
-
-    private PluginClassLoader createClassLoader(List<Artifact> artifacts, String name)
-            throws IOException
-    {
-        log.debug("Classpath for %s:", name);
-        List<URL> urls = new ArrayList<>();
-        for (Artifact artifact : sortedArtifacts(artifacts)) {
-            if (artifact.getFile() == null) {
-                throw new RuntimeException("Could not resolve artifact: " + artifact);
-            }
-            File file = artifact.getFile().getCanonicalFile();
-            log.debug("    %s", file);
-            urls.add(file.toURI().toURL());
-        }
-        return createClassLoader(urls);
-    }
-
     private PluginClassLoader createClassLoader(List<URL> urls)
     {
         ClassLoader parent = getClass().getClassLoader();
         return new PluginClassLoader(urls, parent, SPI_PACKAGES);
     }
 
-    private static List<File> listFiles(File installedPluginsDir)
+    public interface PluginsProvider
     {
-        if (installedPluginsDir != null && installedPluginsDir.isDirectory()) {
-            File[] files = installedPluginsDir.listFiles();
-            if (files != null) {
-                Arrays.sort(files);
-                return ImmutableList.copyOf(files);
-            }
+        void loadPlugins(Loader loader, ClassLoaderFactory createClassLoader);
+
+        interface Loader
+        {
+            void load(String plugin, Supplier<PluginClassLoader> getClassLoader);
```

Comment:
We pass `ClassLoaderFactory` into `PluginsProvider`.
`ClassLoaderFactory` seems the only source of `PluginClassLoader`, which we take here.

Perhaps replace `Supplier<PluginClassLoader> getClassLoader` with `List<URL> pluginClasspath`, to make the control flow simpler?

We can re-introduce this indirection later when needed.

Reply:
The control flow is inverted in this way so that the logging can be done consistently in `PluginManager`. We need `ClassLoaderFactory` in `DevelopmentPluginsProvider` because it needs the class loader itself for `PluginDiscovery`, which can write out a new service descriptor and add it to the class loader. I didn't want to make any changes to that code, since the goal of this PR is simply to move this development code out.

Reply:
> We can re-introduce this indirection later when needed.

so it is needed. thanks for clarifying.